datafusion/core/src/physical_plan/sorts/sort.rs

```diff
@@ -253,11 +256,19 @@ impl ExternalSorter {
     /// Appends an unsorted [`RecordBatch`] to `in_mem_batches`
     ///
     /// Updates memory usage metrics, and possibly triggers spilling to disk
-    async fn insert_batch(&mut self, input: RecordBatch) -> Result<()> {
+    async fn insert_batch(&mut self, mut input: RecordBatch) -> Result<()> {
         if input.num_rows() == 0 {
             return Ok(());
         }
 
+        let mut batch_sorted = false;
+        if self.fetch.map_or(false, |f| f < input.num_rows()) {
```

Comment:
I was also thinking something similar but along the lines of `f < input.num_rows() && f <= 100`, so that we effectively have a hard cur-off for eager sorting after 100 rows.